pkg/master/controller.go

```diff
@@ -150,7 +150,22 @@ func (c *Controller) Stop() {
 		c.runner.Stop()
 	}
 	endpointPorts := createEndpointPortSpec(c.PublicServicePort, "https", c.ExtraEndpointPorts)
-	c.EndpointReconciler.StopReconciling("kubernetes", c.PublicIP, endpointPorts)
+	finishedReconciling := make(chan struct{})
+	go func() {
+		defer close(finishedReconciling)
+		glog.Infof("Shutting down kubernetes service endpoint reconciler")
+		if err := c.EndpointReconciler.StopReconciling("kubernetes", c.PublicIP, endpointPorts); err != nil {
+			glog.Error(err)
+		}
+	}()
+
+	select {
+	case <-finishedReconciling:
+		// done
+	case <-time.After(2 * c.EndpointInterval):
```

Comment:
not sure how 2 is chosen here.

Reply:
purpose of the patch is to eventually give up on writing reconcile state when we are about to shut down kubelet.  Waiting 2x of normal time is gives reconciler enough time to sync state, but also doesn't make us wait too long if write to etcd can't happen.

Reply:
After waiting 2x the normal time we can be pretty sure we've lost the lease and there's not much point in waiting longer.